validation/hooks_stdin.go

```diff
@@ -0,0 +1,126 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"reflect"
+	"time"
+
+	tap "github.com/mndrix/tap-go"
+	rspecs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/opencontainers/runtime-tools/specerror"
+	"github.com/opencontainers/runtime-tools/validation/util"
+	uuid "github.com/satori/go.uuid"
+)
+
+func stdinStateCheck(outputDir, hookName string, expectedState rspecs.State) error {
+	var state rspecs.State
+	data, err := ioutil.ReadFile(filepath.Join(outputDir, hookName))
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal(data, &state)
+	if err != nil {
+		return err
+	}
+
+	if state.Version != expectedState.Version {
+		return fmt.Errorf("wrong OCI specification version %q in the stdin of %s hook, expected %q", state.Version, hookName, expectedState.Version)
+	}
+
+	if state.ID != expectedState.ID {
+		return fmt.Errorf("wrong container ID %q in the stdin of %s hook, expected %q", state.ID, hookName, expectedState.ID)
+	}
+
+	if state.Bundle != expectedState.Bundle {
+		return fmt.Errorf("wrong bundle directory %q in the stdin of %s hook, expected %q", state.Bundle, hookName, expectedState.Bundle)
+	}
+
+	if hookName != "poststop" && state.Pid != expectedState.Pid {
+		return fmt.Errorf("wrong container process ID %q in the stdin of %s hook, expected %q", state.Version, hookName, expectedState.Version)
+	}
+
+	if !reflect.DeepEqual(state.Annotations, expectedState.Annotations) {
+		return fmt.Errorf("wrong annotations \"%v\" in the stdin of %s hook, expected \"%v\"", state.Annotations, hookName, state.Annotations)
+	}
+	return nil
+}
+
+func main() {
+	t := tap.New()
+	t.Header(0)
+
+	bundleDir, err := util.PrepareBundle()
+	if err != nil {
+		util.Fatal(err)
+	}
+	containerID := uuid.NewV4().String()
+	defer os.RemoveAll(bundleDir)
+
+	var containerPid int
+
+	annotationKey := "org.opencontainers.runtime-tools"
+	annotationValue := "hook stdin test"
+	g := util.GetDefaultGenerator()
+	outputDir := filepath.Join(bundleDir, g.Spec().Root.Path)
+	g.AddAnnotation(annotationKey, annotationValue)
+	g.AddPreStartHook(rspecs.Hook{
+		Path: filepath.Join(bundleDir, g.Spec().Root.Path, "/bin/sh"),
+		Args: []string{
+			"sh", "-c", fmt.Sprintf("timeout -t 1 cat > %s", filepath.Join(outputDir, "prestart")),
```

Comment:
Where is your `timeout` from.  It's [not in POSIX][1], and [coreutils uses `-k` / `--kill-after`][2], not `-t`.

[1]: http://pubs.opengroup.org/onlinepubs/9699919799/idx/utilities.html
[2]: http://man7.org/linux/man-pages/man1/timeout.1.html

Reply:
It is in the busybox 1.28, in the new rootfs-amd64.tar.gz.

Reply:
> It is in the busybox 1.28, in the new rootfs-amd64.tar.gz.

The lack of standardization is unfortunate, because hooks are called from the runtime namespace, so the version in the container doesn't matter.  We could work around that with `{bundle-path}/rootfs/bin/timeout` here, but it's probably better to set `Timeout` in the hook structure and skip the `timeout` command.

Reply:
thanks, the timeout in hooks is much more rational.